### ISSUE TITLE
parallel-group-params-mismatch caused by browser

### ISSUE THREAD

**Issue by MarekSuchanek** [cypress-io/github-action#121]
We use parallelized actions with `browser: chrome`, it already happened twice (in ~40 runs) that Chrome version caused fail on "parallel-group-params-mismatch". First job used the older version (e.g. "80.0.3987.132") and others newer (e.g. "80.0.3987.149").

Restarting the job of course does not solve anything as the ID of the action is still the same and with the older version of Chrome. Only solution is to trigger a new run.

Is there please any suggestion on how to solve this? (We wouldn't mind if matching would be just on major version of Chrome, but sadly I did not figure how to do that.)

**Comment by jennifer-shehane** [comment]
What is the exact error you are seeing?

I'm not sure I understand why one job would run on a different version of Chrome. Are you using @latest for Chrome version or something in your setup? You may want to stick to a specific Chrome version so it doesn't update mid-run if that's what's happening - "80.0.3987.132" is not the same version as "80.0.3987.149" and could have distinguishable differences in behavior even though it's not 'major'.

**Comment by MarekSuchanek** [comment]
See the error bellow (or is it possible to see here: https://github.com/ds-wizard/dsw-e2e-tests/runs/549243589).

The GitHub action [allows to specify](https://github.com/cypress-io/github-action/blob/master/action.yml) only the browser name... Is there some way how to specify a version? But the error message says that "browserVersion (major)" must be matching.

For what should match (actions run and cypress.io run):
- specs = in both cases our 32 specs
- osName = in both cases `linux`
- osVersion= in both cases `Ubuntu Linux - 16.04`
- browserName= in both cases `Chrome`
- browserVersion (major) = in both cases `80`
- browserVersion = `80.0.3987.132` in the first action and cypress.io, `80.0.3987.149` in remaining actions

My guess is that chrome gets updated during the run, some parallel jobs will report the older, some the newer.

```
You passed the --parallel flag, but we do not parallelize tests across different environments.

This machine is sending different environment parameters than the first machine that started this parallel run.

The existing run is: https://dashboard.cypress.io/projects/dx9eqg/runs/35

In order to run in parallel mode each machine must send identical environment parameters such as:

- specs
- osName
- osVersion
- browserName
- browserVersion (major)

This machine sent the following parameters:

{
  "osName": "linux",
  "osVersion": "Ubuntu Linux - 16.04",
  "browserName": "Chrome",
  "browserVersion": "80.0.3987.149",
  "specs": [...],
}

https://on.cypress.io/parallel-group-params-mismatch
Error: The process '/usr/local/bin/npx' failed with exit code 1
    at ExecState._setResult (/home/runner/work/_actions/cypress-io/github-action/v1/dist/index.js:1188:25)
    at ExecState.CheckComplete (/home/runner/work/_actions/cypress-io/github-action/v1/dist/index.js:1171:18)
    at ChildProcess.<anonymous> (/home/runner/work/_actions/cypress-io/github-action/v1/dist/index.js:1071:27)
    at ChildProcess.emit (events.js:210:5)
    at maybeClose (internal/child_process.js:1021:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:283:5)
##[error]The process '/usr/local/bin/npx' failed with exit code 1

**Comment by jennifer-shehane** [comment]
Well, the error message does say that it only compares browserVersion (major). Let's see if this is true!

Internal message error from API also says the same:

```
Different run parameters cannot be used for a parallelized group. All parameters such as the specs, osName, osVersion, browserName, and major browserVersion must be the same.
```

Basically we generate a groupId by joining group name, os name, browser name, major version, and a hash generated from the joined specfile names after sorting - so something like `english-linux-Chrome-80-1sjdf90ks`. Then we compare the uniqueness of the groupId. 

Major version is pulled from `const major = (version) => version.split('.')[0]`. I actually don't even see us using the osVersion as indicated in the error message 

For internal reference: https://github.com/cypress-io/cypress-services/blob/develop/packages/api/lib/util/group_id.ts#L23:L23

Are you sure the specs array is *exactly* the same?

I feel like this error could be more helpful though, so I'll see if I can follow up on highlighting more helpful info here. 

(Also not sure this is related to github-actions exactly, so may move this into main repo depending.)

**Comment by jennifer-shehane** [comment]
Actually, looking more, I'm not so sure this is how the comparison is later made. 

https://github.com/cypress-io/cypress-services/blob/develop/packages/api/lib/models/instance.ts#L498:L498

**Comment by MarekSuchanek** [comment]
Yes, the specs are the same, it ends up that from 4 parallel jobs, all specs are done by the first one and others fail...

Thanks for looking into this!

**Comment by jennifer-shehane** [comment]
@MarekSuchanek I've confirmed with the engineering team that we are comparing an exact browser version match instead of just a 'major' version comparison as the error message implies. It should compare major browser version, so this is a bug.

**Comment by MarekSuchanek** [comment]
@jennifer-shehane Thank you!

**Comment by jennifer-shehane** [comment]
There is a merged PR to fix this, so this will be fixed in the next release from our API team. I'll update this message as soon as it's released.

**Comment by jennifer-shehane** [comment]
@MarekSuchanek A fix for this has been deployed and this issue should be resolved. 
If you encounter any similar issues, please open an issue on our main project, you can open it [here](https://github.com/cypress-io/cypress/issues/new).

**Comment by choonkeat** [comment]
Same

> This machine is sending different environment parameters than the first machine that started this parallel run.
> ...
> "browserVersion": "101.0.4951.64"`

whereas the run started by that first machine shows `Chrome 102` on https://dashboard.cypress.io

**Comment by conversaShawn** [comment]
@choonkeat Can you confirm which `cypress/github-action` version number you are using in your `.yml` file?

**Comment by choonkeat** [comment]
@conversaShawn I'm using `cypress-io/github-action@v3.1.0`

**Comment by CharlesStover** [comment]
I'm getting a similar issue, although I'm on `cypress-io/github-action@v2`. The first 2 runs use Edge 106, but the third run uses Edge 107, which throws the `parallel-group-params-mismatch` error.

This is definitely unintuitive, because they're just set to `browser: edge`, and there's no way to specify a version. It's weird to see them automatically choose the wrong version then fail for doing so.

**Comment by gavinhenderson** [comment]
@CharlesStover I am having the same issue with with Chrome

**Comment by chinanderm** [comment]
Happening to me as well right now. Bouncing between Chrome 106 and 107.

**Comment by aminao-transformdata** [comment]
cypress run --record --parallel --group e2e-rendring-smoke --config-file ./cypress.config.js --browser chrome --spec '**/*rendering/**/*,**/rendering*' --tag 'e2e,rendering-smoke,frontend'

It looks like this is your first time using Cypress: 11.2.0

[STARTED] Task without title.
[SUCCESS] Task without title.

Opening Cypress...
You passed the --parallel flag, but we do not parallelize tests across different environments.

This machine is sending different environment parameters than the first machine that started this parallel run.

The existing run is: https://cloud.cypress.io/projects/bjec36/runs

In order to run in parallel mode each machine must send identical environment parameters such as:

 - specs
 - osName
 - osVersion
 - browserName
 - browserVersion (major)

This machine sent the following parameters:

{
  "osName": "darwin",
  "osVersion": "20.6.0",
  "browserName": "Chrome",
  "browserVersion": "107.0.5[30](https://github.com/transform-data/web/actions/runs/3641129875/jobs/6146802099#step:4:31)4.121",
  "specs": [
    "e2e/ApiKeys/rendering.js",
    "e2e/BoardsPage/rendering.js",
    "e2e/HomePage/rendering.js",
    "e2e/InstallPage/rendering.js",
    "e2e/MetricSearchPage/rendering.js",
    "e2e/ModelPage/rendering.js",
    "e2e/NavBar/rendering.js",
    "e2e/Security/rendering.ts",
    "e2e/SettingsPage/rendering.js",
    "e2e/SettingsPanel/rendering.js",
    "e2e/AdminSettingsPage/UserManagement/rendering.js",
    "e2e/ApiExplorerPage/OrgUsers/rendering.js",
    "e2e/ApiExplorerPage/TransformEmployees/rendering.js",
    "e2e/MetricPage/Annotations/rendering.js",
    "e2e/MetricPage/Header/rendering.js",
    "e2e/MetricPage/MetricChart/rendering.js",
    "e2e/MetricPage/MetricDefinition/rendering.js",
    "e2e/MetricPage/Questions/rendering.js",
    "e2e/MetricPage/RightPanel/rendering.js",
    "e2e/QueryLogs/QueryLogsListPage/rendering.js",
    "e2e/Teams/TeamSettings/rendering.js",
    "e2e/MetricPage/RightPanel/MetricOwner/rendering.js",
    "e2e/MetricPage/RightPanel/NonMetricOwner/rendering.js",
    "e2e/MetricPage/RightPanel/OrgAdmin/rendering.js"
  ]
}

**Comment by aminao-transformdata** [comment]
will this be solved?????

**Comment by kolaente** [comment]
This appears to still be an issue, 5 years after this issue was originally opened.

Why is it not possible to just pin the browser version? That would solve this.

As a workaround, I've adjusted my pipeline to this:

```yaml
  define-frontend-e2e-batches:
    runs-on: ubuntu-latest
    outputs: 
      batches: ${{ steps.get-batches.outputs.batches }}
    steps:
      - uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4
      - name: Get spec batches
        id: get-batches
        run: echo "batches=$(find frontend/cypress/e2e -type f -name '*.spec.ts' -printf "%P\n" | jq -R . | jq -cs .)" >> "$GITHUB_OUTPUT"

  test-frontend-e2e:
    runs-on: ubuntu-latest
    needs:
      - api-build
      - frontend-build
      - define-frontend-e2e-batches
    strategy:
      matrix:
        spec: ${{ fromJSON(needs.define-frontend-e2e-batches.outputs.batches) }}
    steps:
     # ...setup steps omitted
      - uses: cypress-io/github-action@v6
        with:
          working-directory: frontend
          browser: chrome
          record: true
          spec: cypress/e2e/${{ matrix.spec }}
          start: |
            pnpm run preview
          wait-on: http://127.0.0.1:4173,http://127.0.0.1:3456/api/v1/info
          wait-on-timeout: 10
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          CYPRESS_RECORD_KEY: ${{ secrets.CYPRESS_RECORD_KEY }}
          CYPRESS_API_URL: http://127.0.0.1:3456/api/v1
          CYPRESS_TEST_SECRET: averyLongSecretToSe33dtheDB
          CYPRESS_DEFAULT_COMMAND_TIMEOUT: 60000
```

This will run one step per spec file, using a matrix strategy. This seems pretty wasteful in terms of resources, but I've yet to find a better option.

**Comment by MikeMcC399** [comment]
@kolaente 

This is a short-coming of the way that GitHub Actions randomly selects runners. During a rollout the runners can be a mixture of old and new versions.

See the explanation in https://github.com/cypress-io/github-action/blob/master/README.md#parallel and suggestion for a workaround using Docker images

> During staged rollout of a new GitHub-hosted runner version, GitHub may provide a mixture of current and new image versions used by the container matrix. It is recommended to use a [Docker image](https://github.com/cypress-io/github-action/blob/master/README.md#docker-image) in the parallel job run which avoids any Cypress Cloud errors due to browser major version mismatch from the two different image versions. A [Docker image](https://github.com/cypress-io/github-action/blob/master/README.md#docker-image) is not necessary if testing against the default built-in Electron browser because this browser version is fixed by the Cypress version in use and it is unaffected by any GitHub runner image rollout.

If you have a problem implementing the suggestion, please open a new issue.

**Comment by kolaente** [comment]
@MikeMcC399 thanks for pointing that out, seems like a cleaner solution. I was able to implement this with a little hiccup along the way, as my dev api server wouldn't work in the docker container out of the box, but that's solved now (unrelated to cypress).
